man/segyio-catb.1
.TH SEGYIO-CATB 1
.SH NAME
segyio-catb \- Concatenate SEG-Y binary header and print on seismic unix format
.SH SYNPOSIS
.B segyio-catb
[\fIOPTION\fR]...
[\fIFILE\fR]...
.SH DESCRIPTION
.B segyio-catb
Concatenate the SEG-Y binary header in FILEs.

.SH OPTIONS
.TP
.BR \-n ", " \-\-nonzero
only print fields with non-zero values

.TP
.BR \-d ", " \-\-description
print with byte offset and field description

.TP
.BR \-\-version
output version information and exit

.TP
.BR \-\-help
display this help and exit

.SH COPYRIGHT
Copyright © Statoil ASA. License LGPLv3+: GNU LGPL version 3 or later <http://gnu.org/licenses/lgpl.html>.

.PP
This is free software: you are free to change and redistribute it.  There is NO WARRANTY, to the extent permitted by law.

.SH SEE ALSO
segyio-cath(1), segyio-catb(1)


changelog.md
# 1.9.13
* Distribute wheels for Python 3.13
* Support for python 3.6, 3.7, 3.8 has been dropped, as it is end-of-life

# 1.9.12
* Distribute wheels for Python 3.12
* Distribute wheels for Apple ARM

# 1.9.11
* Fixed a bug where segyio used the Delay Recording Time without it's scaling
  factor when setting up the sample offsets on open.

# 1.9.10
* Distribute wheels for Python 3.11

# 1.9.9
* Packaging updates

# 1.9.8
* Packaging updates

# 1.9.7
* Packaging updates

# 1.9.6
* gather[] selects the right offsets
* Add python3.9 packages

# 1.9.5
* The 1.9.5 release drops pre-built packages for Python 2.7.
  It will probably still build fine, but no official wheels
  will be available, as pip drops support.

# 1.9.4
* Use Py_ssize_t for type-length arguments

# 1.9.3
* The group feature has been added to docs
* Some infrastructure and CI improvements

# 1.9.2
* segyio has learned to understand rev2 1-byte header words
* segyio now understand rev2 wide header word replacements
* segyio now detects traces with >65k samples
* Support for int24 formats
* Fixed a bug where f.trace[i,j] returned an array even when i,j are ints
* SEGY_INVALID_SORTING message has been improved
* SEGY_SIGNED_INTEGER_3_BYTE enum alias for symmetry with signed counterpart
* Minor build system and CI improvements

# 1.9.1
* Text headers with null-bytes are no longer truncated
* Slowness on sparse writes is pointed out in documentation
* The python package now respects CMAKE_INSTALL_PREFIX
* Multibuild is now cloned on travis, not a submodule
* tools.wrap now works in python3, by decoding the string
* A few C compiler warnings addressed
* collections.abc is imported a future-proof way in gather.py

# 1.9.0
* segyio has learned to work with with >32k samples per trace
* collections.abc is imported a future-proof way
* Support for python 3.4 has been dropped, as it is end-of-life
* Python 3.8 support

# 1.8.8
* Build and packaging fixes

# 1.8.7
* A new feature, group, has been added with experimental support (hence no
  minor release). This is intended for pre-stack files, and is useful for
  grouping up traces by shot gathers and similar
* Segyio has learned how to read subtraces, f.trace[trno, first:last]
* Source water depth is now correctly a 4-byte word
* Fixed a bug with header-assignments in pre-stack files
* A gallery has been added to the docs, for rich examples
* Various documentation improvements, examples, and spelling corrections
* The python build system is rewritten to use scikit-build
* The exported cmake targets are namespaced with segyio::
* segyio now requires CMake >= 3.5 (from 2.8.12) to build

# 1.8.6
* Fixed a bug where negative short header words on gcc were read wrong

# 1.8.5
* segyio now reads short attributes from little-endian files correctly

# 1.8.4
* The from_array function has been refactored, and speed is improved
* create works for non-4-byte numeric formats
* More numerical formats are accepted: 8-byte/64 bit double-precision IEEE
  float, 1, 2, and 4-byte signed and unsigned integers
* segyio no longer crashes on key misses in f.gather[int, :]
* The intended TypeError is raised when traces are accessed with None, instead
  of a misleading (and wrong) AttributeError
* sample-interval has gotten some new tests
* Minor behavorial improvements to C++ test-config
* Some stack-related bug fixes in the test suite
* The readme has seen some minor updates

# 1.8.3
* Check endianness variable properly

# 1.8.2
* Pass HOST_BIG_ENDIAN as preprocessor directive, not compiler option

# 1.8.1
* Fixed a docstring formatting issue
* Fixed a race condition in test suite

# 1.8.0
* segyio has learned the seismic unix file format, and can read and write SU
  files. This feature is found in the segyio.su module
* segyio supports least-significant-byte (LSB/little-endian) first formatted
  files
* Fixed a bug that caused writing headers in non-linear mode very slow,
  and wrong under some conditions
* segyio also considers files with descending line/offset numbers sorted, not
  just ascending
* The python file handle has learned the interpret method, which allows
  specifying structure directly, rather than inferring it from file metadata.
  This is particularly useful for files with structure, but broken headers
* A new family of from_array functions are added to the tools module,
  intended for quick-and-easy storing a 2/3/4D volume on disk
* Sorting defaults to inline in segyio.create if nothing is specified
* Text.__setitem__ used a wrong variable and didn't really work, but is now
  fixed
* Internal imports are tidied up
* The test suite in both Python and C++ have seen some minor improvements
* An experimental C++ interface is available for source builds, but is not
  considered stable - internals and interface can change with no prior notice.
  The experimental interface is enabled when segyio is built with
  -DEXPERIMENTAL=ON

# 1.7.1
* Fixed an issue where writing traces on memory-mapped files was a no-op
* Depth slices dimensions are always in fast-by-slow
* catb/catr has some new output formats
* Some docs entries updated and fixed

# 1.7.0
* segyio has learned to open non-ascii file paths on Windows
* Fixed an issue where segyio-catr would sometimes print the same trace twice
* segyio no longer incorrectly considers files where the
  inline/crossline/offset triple doesn't change sorted
* The test suite has seen some overhaul, making test cases more focused on the
  actual thing to test
* Checking sorting won't silently fail when the offset-word is broken
* `f.text` is now a Sequence, meaning it has slicing and other common array
  operations
* The makefiles no longer build both static and shared libraries, but rather
  respect the BUILD_SHARED_LIBS variable
* Python 3.7 support
* Various refactoring and internal improvements

# 1.6.0
* segyio has learned to handle int16 and int32 sample formats
* segyio has learned to write any array_likes with any float-convertible

applications/segyio-catb.c
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <getopt.h>
#include <unistd.h>

#include "apputils.h"
#include <segyio/segy.h>

static int printhelp(void){
    puts( "Usage: segyio-catb [OPTION]... [FILE]...\n"
          "Concatenate the binary header from FILE(s) to seismic unix "
          "output.\n"
          "\n"
          "-n,  --nonzero       only print fields with non-zero values\n"
          "-d,  --description   print with byte offset and field description\n"
          "     --version       output version information and exit\n"
          "     --help          display this help and exit\n"
          "\n"
        );
    return 0;
}

static int get_binary_value( const char* binheader, int bfield ){
    int32_t f;
    segy_get_bfield( binheader, bfield, &f );

    /*
     * convert cannot-be-negative values to unsigned int, as mandated by SEGY-Y
     * rev2
     */
    switch (bfield) {
        case SEGY_BIN_SAMPLES:
        case SEGY_BIN_SAMPLES_ORIG:
            f = (int32_t)((uint16_t)(f));
            break;
    }
    return f;
}

struct options {
    int version;
    int help;
    int nonzero;
    int description;
    const char* errmsg;
};

static struct options parse_options( int argc, char** argv ){
    struct options opts;
    opts.nonzero = 0;
    opts.description = 0;
    opts.version = 0, opts.help = 0;
    opts.errmsg = NULL;
    
    static struct option long_options[] = {
        {"version",         no_argument,    0,    'V'},
        {"help",            no_argument,    0,    'h'},
        {"description",     no_argument,    0,    'd'},
        {"nonzero",         no_argument,    0,    'n'},
        {0, 0, 0, 0}
    };

    opterr = 1;
     
    while( true ){

        int option_index = 0;
        int c = getopt_long( argc, argv, "nd",
                            long_options, &option_index);

        if ( c == -1 ) break;
        
        switch( c ){
            case  0: break;
            case 'h': opts.help = 1;    return opts;
            case 'V': opts.version = 1; return opts;
            case 'd': opts.description = 1; break;
            case 'n': opts.nonzero = 1; break;
            default:
                 opts.help = 1;
                 opts.errmsg = "";
                 return opts;
        }
    }
    return opts;
}

int main( int argc, char** argv ){

   static const char* su[ 30 ] = {
        "jobid",
        "lino",
        "reno",
        "ntrpr",
        "nart",
        "hdt",
        "dto",
        "hns",
        "nso",
        "format",
        "fold",
        "tsort",
        "vscode",
        "hsfs",
        "hsfe",
        "hslen",
        "hstyp",
        "schn",
        "hstas",
        "hstae",
        "htatyp",
        "hcorr",
        "bgrcv",
        "rcvm",
        "mfeet",
        "polyt",
        "vpol",
        "rev",
        "trflag",
        "exth"
    };

    static const char* su_desc[ 30 ] = {
        "Job identification number"                                         ,
        "Line number"                                                       ,
        "Reel number"                                                       ,
        "Number of data traces per ensemble"                                ,
        "Number of auxiliary traces per ensemble"                           ,
        "Sample interval in microseconds (μs)"                              ,
        "Sample interval in microseconds (μs) of original field recording"  ,
        "Number of samples per data trace"                                  ,
        "Number of samples per data trace for original field recording"     ,
        "Data sample format code"                                           ,
        "Ensemble fold"                                                     ,
        "Trace sorting code"                                                ,
        "Vertical sum code"                                                 ,
        "Sweep frequency at start (Hz)"                                     ,
        "Sweep frequency at end (Hz)"                                       ,
        "Sweep length (ms)"                                                 ,
        "Sweep type code"                                                   ,
        "Trace number of sweep channel"                                     ,
        "Sweep trace taper length in milliseconds at start if tapered"      ,
        "Sweep trace taper length in milliseconds at end"                   ,
        "Taper type"                                                        ,
        "Correlated data traces"                                            ,
        "Binary gain recovered"                                             ,
        "Amplitude recovery method"                                         ,
        "Measurement system"                                                ,
        "Impulse signal polarity"                                           ,
        "Vibratory polarity code"                                           ,
        "SEG Y Format Revision Number"                                      ,
        "Fixed length trace flag"                                           ,
        "Number of 3200-byte, Extended Textual File Headers"
    };

    static const int bfield_value[ 30 ] = {
        SEGY_BIN_JOB_ID,
        SEGY_BIN_LINE_NUMBER,
        SEGY_BIN_REEL_NUMBER,
        SEGY_BIN_TRACES,
        SEGY_BIN_AUX_TRACES,
        SEGY_BIN_INTERVAL,
        SEGY_BIN_INTERVAL_ORIG,
        SEGY_BIN_SAMPLES,
        SEGY_BIN_SAMPLES_ORIG,
        SEGY_BIN_FORMAT,
        SEGY_BIN_ENSEMBLE_FOLD,
        SEGY_BIN_SORTING_CODE,
        SEGY_BIN_VERTICAL_SUM,
        SEGY_BIN_SWEEP_FREQ_START,
        SEGY_BIN_SWEEP_FREQ_END,
        SEGY_BIN_SWEEP_LENGTH,
        SEGY_BIN_SWEEP,
        SEGY_BIN_SWEEP_CHANNEL,
        SEGY_BIN_SWEEP_TAPER_START,
        SEGY_BIN_SWEEP_TAPER_END,
        SEGY_BIN_TAPER,
        SEGY_BIN_CORRELATED_TRACES,
        SEGY_BIN_BIN_GAIN_RECOVERY,
        SEGY_BIN_AMPLITUDE_RECOVERY,
        SEGY_BIN_MEASUREMENT_SYSTEM,
        SEGY_BIN_IMPULSE_POLARITY,
        SEGY_BIN_VIBRATORY_POLARITY,
        SEGY_BIN_SEGY_REVISION,
        SEGY_BIN_TRACE_FLAG,
        SEGY_BIN_EXT_HEADERS
    };
    
    if( argc == 1 ){
         int err = errmsg(2, "Missing argument\n");
         printhelp();
         return err;
    }

    struct options opts = parse_options( argc, argv );
   
    if( opts.help )    return printhelp();
    if( opts.version ) return printversion( "segyio-catb" );

    for( int i = optind; i < argc; ++i ){
        segy_file* fp = segy_open( argv[ i ], "r" );

        if( !fp ) return errmsg(opterr, "No such file or directory");

        char binheader[ SEGY_BINARY_HEADER_SIZE ];
        int err = segy_binheader( fp, binheader );        

        if( err ) return errmsg(opterr, "Unable to read binary header"); 

        for( int c = 0; c < 30; ++c ){
            int field = get_binary_value( binheader, bfield_value[ c ] );
            if( opts.nonzero && !field) continue;

            if( opts.description ) {
                int byte_offset = (bfield_value[ c ] - SEGY_TEXT_HEADER_SIZE);
                printf( "%s\t%d\t%d\t%s\n",
                    su[ c ],
                    field,
                    byte_offset,
                    su_desc[ c ] );
            }
            else
                printf( "%s\t%d\n", su[ c ], field );
        }
        segy_close( fp );
    }
    return 0;
}
